### DISCUSSION TITLE
How do I return string from LineEdit

### DISCUSSION THREAD

**Discussion by dcpri** [slint-ui/slint#572] [Q&A] [answered]
Hi. Any help in passing user input from LineEdit to the calling function? Here is my extracted code (which won't compile because of the SharedString)  :

            LineEdit {
                callback familyname_added(string);
                placeholder-text: "Enter the family name";
                accepted(text) => {
                    familyname_added(text);
                }
            }

pub fn individuals_add() {
    let ui = AppWindow::new();
    let mut my_familyname = String::new;

    ui.on_familyname_added(move |text| {
       	my_familyname = text.to_string();                  // "text" is a SharedString, not a string.    to-string() doesn't work
    });

    ui.run();
}

**Comment by ogoffart** [comment, marked as answer]
SharedString implements the  Rust `Display` trait, so to_string() should work.
What error do you get?

**Reply by dcpri** [reply]
error[E0308]: mismatched types
  --> src/individuals/individuals_add.rs:13:25
   |
13 |         my_familyname = text;
   |                         ^^^^ expected fn item, found struct `SharedString`
   |
   = note: expected fn item `fn() -> String {String::new}`
               found struct `SharedString`

**Reply by ogoffart** [reply]
```
let mut my_familyname = String::new;
```
Should be
```
let mut my_familyname = String::new();
```
It is missing parentheses.

**Reply by dcpri** [reply]
Sorry, I must have deleted them at some point. I have replaced them, but still can't compile. Error is now 
 
error[E0594]: cannot assign to `my_familyname`, as it is a captured variable in a `Fn` closure
  --> src/individuals/individuals_add.rs:14:9
   |
14 |         my_familyname = text.to_string();
   |         ^^^^^^^^^^^^^ cannot assign

I am new to Rust and I am definitely floundering. Sorry.

**Reply by ogoffart** [reply]
Ah right, that's because the `my_familyname` lives on the stack, but the handler needs a `'static` closure

The solution here is to put the data on the heap on a `Rc<RefCell>`

You probably want something like that:

```rust
fn main() {
    sixtyfps::sixtyfps! {
        import {LineEdit} from "sixtyfps_widgets.60";
        AppWindow := Window {
            callback familyname_added(string);
            LineEdit {
                width: 100%;
                height: 100%;
                placeholder_text: "Enter the family name";
                accepted(text) => {
                    familyname_added(text);
                }
            }
        }
    };

    let ui = AppWindow::new();

    #[derive(Default)]
    struct MyData {
        my_familyname: String, // you would as well use a SharedString to save allocations, but that's up to you
    }
    let my_data = std::rc::Rc::new(std::cell::RefCell::new(MyData::default()));

    ui.on_familyname_added({
        let my_data = my_data.clone();
        move |text| {
            my_data.borrow_mut().my_familyname = text.to_string();
            println!("Callback: {}", my_data.borrow().my_familyname);
        }
    });

    ui.run();

    println!("Window Closed: {}", my_data.borrow().my_familyname);
}
```

I hope this helps

**Reply by dcpri** [reply]
Many thanks. My program compiles with your code but it is not yet doing the callback. I have my code over several files which is probably the problem. Everything displays fine so I will keep working on the callback. Thanks for your time.
